### PR TITLE
Add tiled matmul example

### PR DESCRIPTION
**Pull request by JonasCrols**
Added support for snax-stream-gemm, used by adding a matmul that is tiled by using rectangular non-accumulating tiles.

### CODE REVIEW

File: runtime/Makefile.rules [hunk 1]
```diff
@@ -125,6 +125,7 @@ MLIROPTFLAGS += --llvm-request-c-wrappers
 MLIROPTFLAGS += --expand-strided-metadata
 MLIROPTFLAGS += --convert-index-to-llvm=index-bitwidth=32
 MLIROPTFLAGS += --convert-cf-to-llvm=index-bitwidth=32
+MLIROPTFLAGS += --lower-affine
```
Comment (JosseVanDelm) [inline runtime/Makefile.rules:128]: Why do we need this?

**Review by JosseVanDelm — COMMENTED** [review]
Thanks for your work @JonasCrols, could you still change the loops to be non-inter-accumulating and also make your PR description a bit more descriptive?

Also, why do we suddenly need `lower-affine?` Thanks!

File: runtime/Makefile.rules (hunk 1 above)
Comment (jorendumoulin) [reply runtime/Makefile.rules:128]: If i remember correctly, there were some leftover affine.apply ops which were throwing errors down the line. I'm not sure where they appear and why the previous --lower-affine does not catch them, but adding this pass resolves the errors.

**Review by JosseVanDelm — APPROVED** [review]
Please change compilation pass order and remove non-functional example :heart:

File: tests/filecheck/transforms/set-memory-layout.mlir [hunk 2]
```diff
@@ -5,7 +5,7 @@
   ^0(%arg0 : memref<?x128xi8, 1 : i32>, %arg1 : memref<128x128xi8, 1 : i32>, %arg2 : memref<?x128xi32, 1 : i32>):
     %0 = "arith.constant"() <{"value" = 1 : i32}> : () -> i32
     %1 = "arith.constant"() <{"value" = 1 : i32}> : () -> i32
-    "linalg.generic"(%arg0, %arg1, %0, %1, %arg2) <{"indexing_maps" = [affine_map<(d0, d1, d2) -> (d0, d2)>, affine_map<(d0, d1, d2) -> (d2, d1)>, affine_map<(d0, d1, d2) -> ()>, affine_map<(d0, d1, d2) -> ()>, affine_map<(d0, d1, d2) -> (d0, d1)>], "iterator_types" = [#linalg.iterator_type<parallel>, #linalg.iterator_type<parallel>, #linalg.iterator_type<reduction>], "operandSegmentSizes" = array<i32: 4, 1>, "library_call" = "snax_gemm"}> ({
```
Comment (JosseVanDelm) [inline tests/filecheck/transforms/set-memory-layout.mlir:8]: This change seems unnecessary

File: tests/filecheck/transforms/dispatch_regions.mlir [hunk 3]
```diff
@@ -113,4 +113,43 @@
...
+    } ins(%0, %1 : memref<64xi32>, memref<64xi32>) outs(%2 : memref<64xi32>) {
+    ^bb0(%arg0: i32, %arg1: i32, %arg2: i32):
+      %3 = arith.muli %arg0, %arg1 : i32
+      linalg.yield %3 : i32
+  }
+    "func.return"() : () -> ()
+
+  }) : () -> ()
+}) : () -> ()
+
```
Comment (JosseVanDelm) [inline tests/filecheck/transforms/dispatch_regions.mlir:135]: how is this testing something different from the above test?

File: kernels/simple_matmul/main.c [hunk 4]
```diff
@@ -21,8 +21,9 @@
  * /target/snitch_cluster/sw/snax/mac/include"
  *
  * */
-#include "snax-gemm-lib.h"
```
Comment (JosseVanDelm) [inline kernels/simple_matmul/main.c:24]: Can we not include these routines now, instead of copy-pasting them?

File: tests/filecheck/transforms/dispatch_kernels.mlir [hunk 5]
```diff
@@ -44,7 +44,7 @@
 //CHECK-NEXT:   ^0(%arg0 : memref<?x128xi8>, %arg1 : memref<128x128xi8>, %arg2 : memref<?x128xi32>):
 //CHECK-NEXT:     %0 = "arith.constant"() <{"value" = 1 : i32}> : () -> i32
 //CHECK-NEXT:     %1 = "arith.constant"() <{"value" = 1 : i32}> : () -> i32
-//CHECK-NEXT:     "linalg.generic"(%arg0, %arg1, %0, %1, %arg2) <{"indexing_maps" = [affine_map<(d0, d1, d2) -> (d0, d2)>, affine_map<(d0, d1, d2) -> (d2, d1)>, affine_map<(d0, d1, d2) -> ()>, affine_map<(d0, d1, d2) -> ()>, affine_map<(d0, d1, d2) -> (d0, d1)>], "iterator_types" = [#linalg.iterator_type<parallel>, #linalg.iterator_type<parallel>, #linalg.iterator_type<reduction>], "operandSegmentSizes" = array<i32: 4, 1>, "library_call" = "snax_gemm"}> ({
+//CHECK-NEXT:     "linalg.generic"(%arg0, %arg1, %0, %1, %arg2) <{"indexing_maps" = [affine_map<(d0, d1, d2) -> (d0, d2)>, affine_map<(d0, d1, d2) -> (d2, d1)>, affine_map<(d0, d1, d2) -> ()>, affine_map<(d0, d1, d2) -> ()>, affine_map<(d0, d1, d2) -> (d0, d1)>], "iterator_types" = [#linalg.iterator_type<parallel>, #linalg.iterator_type<parallel>, #linalg.iterator_type<reduction>], "operandSegmentSizes" = array<i32: 4, 1>, "library_call" = "snax_qgemm"}> ({
```
Comment (JosseVanDelm) [inline tests/filecheck/transforms/dispatch_kernels.mlir:47]: Why is this changed?

File: kernels/half_tiled_matmul/main.c [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    write_csr(987, 0);
+    write_csr(987, 0);
+    write_csr(981, 0);
+
+    // // implement some artificial delay
+    // for (int i = 0; i < 1000; i++) {{
+    //     asm volatile("nop");
+    // }}
+  }
+}
```
Comment (JosseVanDelm) [inline kernels/half_tiled_matmul/main.c:125]: Same comment here, can we please include these instead of putting them here?

File: .github/workflows/build-run-kernel.yml [hunk 7]
```diff
@@ -10,7 +10,7 @@ jobs:
   build-and-run-kernels:
     runs-on: ubuntu-latest
     container: 
-      image: ghcr.io/kuleuven-micas/snax:v0.1.4
+      image: ghcr.io/kuleuven-micas/snax:v0.1.6
```
Comment (JosseVanDelm) [inline .github/workflows/build-run-kernel.yml:13]: I'd prefer updating all CI routines in one go to use the newest container.
Can you also update the other CI jobs please? Thanks!

File: kernels/half_tiled_matmul/half_tiled_matmul.mlir [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+      %tiled_D = memref.subview %arg2[%iv, %jv][%short_tile_size, %short_tile_size][%c1, %c1] : memref<16x16xi32> to memref<?x?xi32, strided<[16, 1], offset: ?>>
+      
+      // Perform the quantized matrix multiplication on the tiles
+      linalg.quantized_matmul ins(%tiled_A, %tiled_B, %c0_i32, %c0_i32 : memref<?x?xi8, strided<[16, 1], offset: ?>>, memref<?x?xi8, strided<[1, 16], offset: ?>>, i32, i32) outs(%tiled_D : memref<?x?xi32, strided<[16, 1], offset: ?>>)
+      
+    }
+  }
+  
+  return
+}
```
Comment (JosseVanDelm) [inline kernels/half_tiled_matmul/half_tiled_matmul.mlir:23]: add newline

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]
Hey this is nicely coming along!
Hi, sorry but I'm noticing a lot of code copy pasting of which I'm not a really big fan.
I'm also not sure why you changed the library call to `snax_qgemm`, but  feel free to go ahead if this is what you prefer.
I'm also not entirely sure the test you put here tests the previous error?
There are also still a few left-over suggestions from Joren.

File: tests/filecheck/transforms/dispatch_regions.mlir (hunk 3 above)
Comment (JonasCrols) [reply tests/filecheck/transforms/dispatch_regions.mlir:135]: the Dispatch_regions pass fails only when Linalg.generic is used in this way. when it is used in the other way (with quotation marks), the pass always performs correctly. we now test both cases

File: tests/filecheck/transforms/dispatch_kernels.mlir (hunk 5 above)
Comment (JonasCrols) [reply tests/filecheck/transforms/dispatch_kernels.mlir:47]: because the new gemm-call is called snax_qgemm now. is there a way this doesn't need to be changed?

File: kernels/simple_matmul/main.c (hunk 4 above)
Comment (JonasCrols) [reply kernels/simple_matmul/main.c:24]: what exactly would you like included in a separate routine? most of this comes straight out of the template

File: tests/filecheck/transforms/dispatch_regions.mlir (hunk 3 above)
Comment (AntonLydike) [reply tests/filecheck/transforms/dispatch_regions.mlir:135]: That shouldn't be though, the quotation marks just mark generic vs. custom syntax of how the operation is printed. They should represent the exact same operation though...

File: kernels/simple_matmul/main.c [hunk 9]
```diff
@@ -104,4 +104,4 @@ int main() {
       nerr += 1;
   }
   return nerr;
-}
+}
```
Comment (jorendumoulin) [inline kernels/simple_matmul/main.c:107]: add newline

File: runtime/snax-streamer-gemm.rules [hunk 10]
```diff
@@ -0,0 +1,5 @@
+# Specific settings for snax-mac RTL
```
Comment (jorendumoulin) [inline runtime/snax-streamer-gemm.rules:1]:
```suggestion
# Specific settings for snax-streamer-gemm RTL
```

**Review by jorendumoulin — APPROVED** [review]
Perfect, thank you! 
Jus two last final nits!

File: kernels/streamer_matmul/Makefile [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+TESTS =
+TESTS += half_tiled_matmul.x
+TESTS += matmul.x
+
+CFLAGS += -std=gnu11
+CFLAGS += -Wall -Wextra
+
+# Override snax-opt rules to avoid linalg-to-library-call pass
+SNAXOPTACCFLAGS = -p dispatch-kernels,set-memory-space,set-memory-layout,realize-memref-casts,insert-sync-barrier,dispatch-regions,insert-accfg-op{accelerator=snax_gemm},convert-linalg-to-accfg,convert-accfg-to-csr,snax-copy-to-dma,memref-to-snax,snax-to-func,clear-memory-space
+
```
Comment (JosseVanDelm) [inline kernels/streamer_matmul/Makefile:17]: I think this is still running on CPU, can you remove the `*accfg*` passes and insert `convert-linalg-to-library-call` pass?

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]
I think it is not using the accelerator yet

File: kernels/streamer_matmul/Makefile [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+%.acc_dialect.snax-opt.mlir: %.preprocfinal.mlir
+	$(SNAXOPT) $(SNAXOPTACCFLAGS) -o $@ $<
+
```
Comment (JosseVanDelm) [inline kernels/streamer_matmul/Makefile:20]: Remove these lines please

**Review by JosseVanDelm — APPROVED** [review]
Okay gotem

File: kernels/streamer_matmul/Makefile (hunk 11 above)
Comment (JosseVanDelm) [reply kernels/streamer_matmul/Makefile:17]: Turns out these passes are not used anyways :D
